src/cargo/util/dependency_queue.rs

```diff
@@ -82,36 +84,39 @@ impl<N: Hash + Eq + Clone, E: Eq + Hash + Clone, V> DependencyQueue<N, E, V> {
     /// All nodes have been added, calculate some internal metadata and prepare
     /// for `dequeue`.
     pub fn queue_finished(&mut self) {
+        let mut out = HashMap::new();
         for key in self.dep_map.keys() {
-            depth(key, &self.reverse_dep_map, &mut self.depth);
+            depth(key, &self.reverse_dep_map, &mut out);
         }
+        self.priority = out.into_iter().map(|(n, set)| (n, set.len())).collect();
 
-        fn depth<N: Hash + Eq + Clone, E: Hash + Eq + Clone>(
+        fn depth<N: Hash + Eq + Clone + std::fmt::Debug + PartialEq, E: Hash + Eq + Clone>(
             key: &N,
             map: &HashMap<N, HashMap<E, HashSet<N>>>,
-            results: &mut HashMap<N, usize>,
-        ) -> usize {
-            const IN_PROGRESS: usize = !0;
-
-            if let Some(&depth) = results.get(key) {
-                assert_ne!(depth, IN_PROGRESS, "cycle in DependencyQueue");
-                return depth;
+            results: &mut HashMap<N, HashSet<N>>,
+        ) -> HashSet<N> {
+            let in_progress: HashSet<N> = HashSet::new();
+            if let Some(depth) = results.get(key) {
+                assert_ne!(depth, &in_progress, "cycle in DependencyQueue");
+                return depth.clone();
             }
+            results.insert(key.clone(), in_progress);
 
-            results.insert(key.clone(), IN_PROGRESS);
+            let mut set = HashSet::new();
+            set.insert(key.clone());
 
-            let depth = 1 + map
+            for dep in map
                 .get(key)
                 .into_iter()
                 .flat_map(|it| it.values())
                 .flat_map(|set| set)
-                .map(|dep| depth(dep, map, results))
-                .max()
```

Comment:
Instead of storing intermediate hash sets, I think we can basically change this `max()` to `sum()` to get the same effect, right?

Reply:
```
 root
 /   \
a     b
 \   /
   c
````
The sum approach has `c` = 4 the sets approach has `c` = 3. I gut tells me that the set will be a better huristick, but I don't have evidence.

Reply:
d'oh

Reply:
The same two runs get me 246-217s, so on my computer for cargo the set approach is better.